cmd/maya-exporter/app/collector/cstorcollector_test.go

```diff
@@ -0,0 +1,190 @@
+	resp2    = "IOSTATS  IQN=iqn.2017-08.OpenEBS.cstor:vol1 Blockcount=20971520 Blocklength=512 Writes=0 Reads=0 TotalReadBytes=0 TotalWriteBytes=0 Size=10737418240\r\nOK IOSTATS\r\n"
+)
+
+func TestSplit(t *testing.T) {
+	cases := map[string]struct {
+		response string
+		output   map[string]string
+	}{
+		"Response with header": {
+			response: resp1,
+			output: map[string]string{
+				"TotalWriteBytes": "0",
+				"Size":            "10737418240",
+				"IQN":             "iqn.2017-08.OpenEBS.cstor:vol1",
+				"Blockcount":      "20971520",
+				"Blocklength":     "512",
+				"Writes":          "0",
+				"Reads":           "0",
+				"TotalReadBytes":  "0",
+			},
+		},
+		"Response without header": {
+			response: resp2,
+			output: map[string]string{
+				"TotalWriteBytes": "0",
+				"Size":            "10737418240",
+				"IQN":             "iqn.2017-08.OpenEBS.cstor:vol1",
+				"Blockcount":      "20971520",
+				"Blocklength":     "512",
+				"Writes":          "0",
+				"Reads":           "0",
+				"TotalReadBytes":  "0",
+			},
+		},
+	}
+	for name, tt := range cases {
+		t.Run(name, func(t *testing.T) {
+			got := split(tt.response)
+			if len(got) != len(tt.output) {
+				t.Fatalf("split(%v) : expected %v, got %v", tt.response, got, tt.output)
+			}
+		})
+	}
+}
+
+func runFakeUnixServer(t *testing.T, wg *sync.WaitGroup) {
+	go func() {
+		// start the server
+		//var err error
+		t.Log("server started")
+		listener, err := net.Listen("unix", "@anysocketpath")
+		wg.Done()
+		if err != nil {
+			t.Fatal(err)
+		}
+		for {
+			fd, err := listener.Accept()
+			if err != nil {
+				t.Fatal("Accept error: ", err)
+			}
+			go sendFakeResponse(t, fd)
+		}
+	}()
+}
+
+func sendFakeResponse(t *testing.T, c net.Conn) {
+	for {
+		buf := make([]byte, 512)
+		_, err := c.Read(buf)
+		if err != nil {
+			return
+		}
+
+		data := resp1
+		t.Logf("Received: %v", data)
+		_, err = c.Write([]byte(data))
+		if err != nil {
+			panic("Write: " + err.Error())
+		}
+	}
+}
+
+// TestCollector tests cstorcollector.go
+func TestCstorCollector(t *testing.T) {
+
+	for _, tt := range []struct {
+		input string
+		match []*regexp.Regexp
+	}{
+		// this is the input we are passing for positive testing
+		// match will expect similar output from response.
+		{
+			input: `
+{
+	"stats": {
+		"actual_used": 0,
+		"logical_size": 0,
+		"sector_size": 0,
+		"read_iops": 0,
+		"read_time_per_second": 0,
+		"read_block_count_per_second": ,
+		"write_iops": 0,
+		"write_time_per_second": 0,
+		"write_block_count_per_second": 0,
+		"read_latency": 0,
+		"write_latency": 0,
+		"avg_read_block_count_per_second": 0,
+		"avg_write_block_count_per_second": ,
+		"size_of_volume": 10,
+	}
+}`,
+			// match matches the response with the expected input.
+			match: []*regexp.Regexp{
```

Comment:
I believe the regex elements & stats element need to be at exact index for the test logic to work?
Can this be made simpler. Perhaps a struct to hold related elements together?

Reply:
@AmitKumarDas i tried with other approaches and i find it very difficult to test because we don't get the json data, so here i directly fetch the buffer of bytes -> convert into string and match with the regular expression and this is the only way to improve coverage, i.e, all business logic get covered.

Reply:
I felt an external contributor should understand the business logic by reading the UT (& of-course the src code). We should try to make this a reality.